File: src/Algorithm/Arrays/Longest Consecutive Sequence/lcs.cpp
/*
    Problem Link: https://leetcode.com/problems/longest-consecutive-sequence/
    Given unsorted array, return length of longest consecutive sequence
    Ex. nums = [100,4,200,1,3,2] -> 4, longest is [1,2,3,4]

    Time: O(n)
    Space: O(n)
*/

class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> s;
        for (int i = 0; i < nums.size(); i++) {
            s.insert(nums[i]);
        }
        
        int result = 0;
        
        for (auto it = s.begin(); it != s.end(); it++) {
            int currNum = *it;
            if (s.find(currNum - 1) != s.end()) {
                continue;
            }
            int currLength = 1;
            while (s.find(currNum + 1) != s.end()) {
                currLength++;
                currNum++;
            }
            result = max(result, currLength);
        }
        
        return result;
    }
};

File: src/Algorithm/Arrays/LeetcodeProblems/LongestConsecutiveSequence/lcs.java
/*
    Problem Link: https://leetcode.com/problems/longest-consecutive-sequence/
    Given unsorted array, return length of longest consecutive sequence
    Ex. nums = [100,4,200,1,3,2] -> 4, longest is [1,2,3,4]

    Time: O(n)
    Space: O(n)
*/


class Solution {
    public int longestConsecutive(int[] nums) {
        Set<Integer> num_set = new HashSet<Integer>();
        for (int num : nums) {
            num_set.add(num);
        }

        int longestStreak = 0;

        for (int num : num_set) {
            if (!num_set.contains(num-1)) {
                int currentNum = num;
                int currentStreak = 1;

                while (num_set.contains(currentNum+1)) {
                    currentNum += 1;
                    currentStreak += 1;
                }

                longestStreak = Math.max(longestStreak, currentStreak);
            }
        }

        return longestStreak;
    }
}